File: app/src/main/java/com/elishaazaria/sayboard/data/LocalModel.java
package com.elishaazaria.sayboard.data;

import java.io.Serializable;
import java.util.Locale;
import java.util.Objects;

public class LocalModel implements Serializable {
    public final String path;
    public final Locale locale;
    public final String filename;

    public LocalModel(String path, Locale locale, String filename) {
        this.path = path;
        this.locale = locale;
        this.filename = filename;
    }

    public String serialize() {
        return serialize(this);
    }

    public static String serialize(LocalModel model) {
        return "[path:\"" + encode(model.path) +
                "\", locale:\"" + model.locale +
                "\", name:\"" + encode(model.filename) + "\"]";
    }

    public static LocalModel deserialize(String serialized) {
        throw new RuntimeException(); // TODO: implement
    }

    private static String encode(String s) {
        StringBuilder sb = new StringBuilder();
        char c;
        for (int i = 0; i < s.length(); i++) {
            c = s.charAt(i);
            switch (c) {
                case ',':
                case '"':
                case '\\':
                case ':':
                    sb.append("\\");
                    sb.append(String.format("%02x", (int) c));
                    break;
                default:
                    sb.append(c);
            }
        }
        return sb.toString();
    }

    private static String decode(String s) {
        StringBuilder sb = new StringBuilder();
        char c;
        for (int i = 0; i < s.length(); i++) {
            c = s.charAt(i);
            if (c == '\\') {
                i++;
                sb.append((char) Integer.parseInt(s.substring(i, i + 2)));
                i += 2;
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        LocalModel model = (LocalModel) o;

        if (!Objects.equals(path, model.path)) return false;
        if (!Objects.equals(locale, model.locale)) return false;
        return Objects.equals(filename, model.filename);
    }

    @Override
    public int hashCode() {
        int result = path != null ? path.hashCode() : 0;
        result = 31 * result + (locale != null ? locale.hashCode() : 0);
        result = 31 * result + (filename != null ? filename.hashCode() : 0);
        return result;
    }
}

File: app/src/main/java/com/elishaazaria/sayboard/data/LocalModel.kt
package com.elishaazaria.sayboard.data

import java.io.Serializable
import java.util.*

class LocalModel(val path: String?, val locale: Locale?, val filename: String?) : Serializable {
    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val model = o as LocalModel
        if (path != model.path) return false
        return if (locale != model.locale) false else filename == model.filename
    }

    override fun hashCode(): Int {
        var result = path?.hashCode() ?: 0
        result = 31 * result + (locale?.hashCode() ?: 0)
        result = 31 * result + (filename?.hashCode() ?: 0)
        return result
    }

    companion object {
        fun serialize(model: LocalModel): String {
            return "[path:\"" + encode(model.path) +
                    "\", locale:\"" + model.locale +
                    "\", name:\"" + encode(model.filename) + "\"]"
        }

        fun deserialize(serialized: String?): LocalModel {
            throw RuntimeException() // TODO: implement
        }

        private fun encode(s: String?): String {
            val sb = StringBuilder()
            var c: Char
            for (i in 0 until s!!.length) {
                c = s[i]
                when (c) {
                    ',', '"', '\\', ':' -> {
                        sb.append("\\")
                        sb.append(String.format("%02x", c.code))
                    }
                    else -> sb.append(c)
                }
            }
            return sb.toString()
        }

        private fun decode(s: String): String {
            val sb = StringBuilder()
            var c: Char
            var i = 0
            while (i < s.length) {
                c = s[i]
                if (c == '\\') {
                    i++
                    sb.append(s.substring(i, i + 2).toInt().toChar())
                    i += 2
                } else {
                    sb.append(c)
                }
                i++
            }
            return sb.toString()
        }
    }
}